[foundation 3 of 3] DCOS-14745: Extract foundation

nLight opened the pull request:
ℹ️  License adds 200LOC, don't be intimidated by +333/-134 :)

## Local packages

- Introduced /packages dir for all the local packages we're not ready to completely extract.
As discussed offline with @orlandohohmeier it is considered beneficial since we don't need to mess around with `npm link` nor `npm install` as we change code inside those packages
- Rewired dcos-ui to catch everything from this dir and treat as an npm package while also including them in the `npm test` and `npm run lint` routines to make our life easier

## DCOSStore and ResourceTableUtil

- Moved from `foundation-ui` to DCOS since they are highly DCOS-UI specific

## PluginSDK

- Removed all the foundation bits from the PluginSDK, plugins now can just use the package
- Exposed DCOSStore and ResourceTableUtil

~Should go together with https://github.com/mesosphere/dcos-ui-plugins-private/pull/565~
~List of PRs: https://github.com/pulls?utf8=%E2%9C%93&q=is%3Aopen+is%3Apr+author%3AnLight+foundation~

🙇

jfurrow commented:
When testing this, I'm receiving an error and a warning. Is this to be expected?

![](https://cl.ly/2G2G3L0r1G40/Screen%20Shot%202017-04-03%20at%201.13.24%20PM.png)

nLight commented:
For some reason webpack wants to include LICENSE in the build. I suspect that's because foundation-ui package sits right there and maybe we need to exclude it somehow from the build.

Do we have webpack Gods on the team?

/cc @orlandohohmeier

mlunoe commented on foundation-ui/src/mount/MountService.js:
```diff
@@ -43,7 +43,7 @@ class MountService extends EventEmitter {
 
     if (typeof component !== 'function') {
-      if (global.__DEV__) {
+      if (process.env.NODE_ENV !== 'production') {
```
❤️

mlunoe commented on src/js/routes/index.js:
```diff
@@ -92,7 +92,7 @@ function getRoutes() {
   // Register packages
   indexRoute.children = indexRoute.children.concat(
-    RoutingService.getDefinition()
+    RoutingService.RoutingService.getDefinition()
```
🤔 Could we not get rid of the duplicated name?

mlunoe commented on src/js/utils/NavigationServiceUtil.js:
```diff
@@ -28,14 +29,14 @@ const NavigationServiceUtil = {
         const primaryPath = `/${path}`;
 
-        NavigationService.registerCategory(category);
+        NavigationService.NavigationService.registerCategory(category);
```
Same here, could we not get rid of that duplicated name?

mlunoe commented on foundation-ui/index.js:
```diff
@@ -1,17 +1,13 @@
-import {MountService} from './mount';
-import Mount from './mount/Mount';
+import RoutingService from './src/routing';
+import NavigationService from './src/navigation';
```
```js
import {NavigationService, EventTypes as NavigationEventTypes} from './src/navigation';
...
module.exports = {
  NavigationService,
  NavigationEventTypes,
};
```
Or something?

nLight replied:
How about:

```
module.exports = {
  navigation: {...},
  routing: {...}
};
```

```
import {navigation: {EventTypes}} from 'foundation-ui';
```

nLight replied:
Technically those are sub-packages...

mlunoe replied:
Yup, let's do it!

mlunoe commented on gulpfile.js:
```diff
@@ -25,7 +25,14 @@ gulp.task('ensureDevProxy', function () {
-    return dependency + '@' + dependencies[dependency];
+    var version = dependencies[dependency];
+
+    if (version.startsWith('./')) {
```
Let's get a comment here. Something like:
`// Module is local, include dependencies from its package.json`

nLight commented:
@mlunoe @orlandohohmeier @jfurrow please give it another shot. I addressed all the comments and improved the original solution with some guidance from Orlando 🙇

mlunoe commented:
This looks really good @nLight! I do get this error a lot navigating around:
![](https://cl.ly/jzUG/Image%202017-04-13%20at%2010.37.48.png)

nLight commented:
@mlunoe good one! Looks like the change made the problem visible, we shouldn't ideally register a mount point in a constructor, rather using the react life cycle methods.

mlunoe commented on plugins/services/src/js/components/modals/NewCreateServiceModalServicePicker.js:
```diff
@@ -49,27 +49,30 @@ function JSONOption({columnClasses, onOptionSelect}) {
+class NewCreateServiceModalServicePicker extends React.Component {
+  componentDidMount() {
+    OPTIONS.forEach(function (component, index) {
+      MountService.MountService.registerComponent(
```
👍

mlunoe commented:
Looking good! Works great!

jfurrow commented:
It works great! Nice work!
